src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -880,17 +878,30 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
   CHECK_EQ(out_attrs->size(), 1U);
   const int in_stype = in_attrs->at(0);
   int& out_stype = out_attrs->at(0);
+  const NormParam& param = nnvm::get<NormParam>(attrs.parsed);
   bool dispatched = false;
+  // l2 norm on a particular axis only supports cpu
+  const bool invalid_ctx = dev_mask != mshadow::cpu::kDevMask;
+  const auto dispatch_ex =
+      invalid_ctx ? DispatchMode::kFComputeFallback : DispatchMode::kFComputeEx;
   if (!dispatched && in_stype == kDefaultStorage) {
     // dns -> dns
     dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
                                      DispatchMode::kFCompute);
   }
-  if (!dispatched && (in_stype == kCSRStorage || in_stype == kRowSparseStorage)) {
-    // csr/rsp -> dns
+  const TShape axis = param.axis.has_value() ? param.axis.value() : TShape();
+  if (!dispatched && (in_stype == kRowSparseStorage || in_stype == kCSRStorage) &&
+      axis.ndim() == 0 && param.ord == 2) {
+    // l2 norm: rsp/csr, axis = () -> dns
     dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
                                      DispatchMode::kFComputeEx);
   }
+  if (!dispatched && in_stype == kCSRStorage && axis.ndim() == 1 && !param.keepdims &&
```

Comment:
the same thing here. checking keepdims here isn't necessary.

Reply:
This is because the cpu csr reduce implementation doesn't support keepdims=True.

Reply:
If we want to support this, csr.norm(axis=0, keepdims=True) should probably return a csr output instead of dense output